File: 5. Sorting/05. RecursiveBubble.cpp
#include <iostream>
#include <vector>
using namespace std;

void recursiveBubbleSort(vector<int>& arr, int n) {

    // Base case
    if (n == 1)
        return;

    // One Bubble Sort pass
    for (int i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            swap(arr[i], arr[i + 1]);
        }
    }

    // Recursive call
    recursiveBubbleSort(arr, n - 1);
}

int main() {

    vector<int> arr = {64, 25, 12, 22, 11};

    recursiveBubbleSort(arr, arr.size());

    for (int x : arr)
        cout << x << " ";

    return 0;
}

File: 5. Sorting/05. RecursiveBubble.py
def recursive_bubble_sort(arr, n):

    # Base case
    if n == 1:
        return

    # One Bubble Sort pass
    for i in range(n - 1):
        if arr[i] > arr[i + 1]:
            arr[i], arr[i + 1] = arr[i + 1], arr[i]

    # Recursive call
    recursive_bubble_sort(arr, n - 1)


# Example
arr = [64, 25, 12, 22, 11]

recursive_bubble_sort(arr, len(arr))

print(arr)